src/server.toit

```diff
@@ -106,26 +106,33 @@ class Server:
     response_writer.write_headers STATUS_SWITCHING_PROTOCOLS
     return WebSocket response_writer.detach --no-client
 
+  // Returns true if the connection was detached, false if it was closed.
   run_connection_ connection/Connection handler/Lambda logger/log.Logger -> bool:
     if call_in_finalizer_: connection.call_in_finalizer_ = call_in_finalizer_
     while true:
       request := null
       with_timeout read_timeout:
         request = connection.read_request
-      if not request: return false
+      if not request: return false  // Timed out waiting for request from client.
       request_logger := logger.with_tag "path" request.path
       request_logger.debug "incoming request"
       writer ::= ResponseWriter connection request request_logger
-      try:
-        handler.call request writer
-      finally: | is_exception exception |
-        // Drain unread content to allow the connection to be reused.
-        if writer.detached_: return true
-        if is_exception:
-          writer.close_on_exception_ "Internal Server error - $exception.value.stringify"
-        else:
-          request.drain
-          writer.close
+      unwind_block := : | exception |
+        // If there's an error we can either send a 500 error message or close
+        // the connection.  This depends on whether we had already sent the
+        // headers - can't send a 500 if we already sent a success header.
+        closed := writer.close_on_exception_ "Internal Server error - $exception"
+        closed   // Unwind if the connection is dead.
+      error := catch --trace --unwind=unwind_block:
+        handler.call request writer  // Calls the block passed to listen.
+      if writer.detached_: return true
+      if foo := request.body.read:
+        // The request (eg. a POST request) was not fully read - should have
+        // been closed and return null from read.
+        closed := writer.close_on_exception_ "Internal Server error - request not fully read"
+        assert: closed
+        throw "request not fully read: $request.path"
```

Comment:
Looks like a weird internal exception string.

Reply:
This is what we send to the browser when the library user's router did not read the whole of the POST data.